### ISSUE TITLE
Native Solaris AIO Reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5538]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5538|
|Type|enhancement|
|Created|2012-03-13 05:42:18Z|

I've recently been made aware of some fun [native interfaces on solaris](http://developers.sun.com/solaris/articles/thread_pools.html) which allow applications to wait on async disk files, async sockets, and maybe other async things equally well.

Such a reactor might also point the way to a portable async filesystem I/O API, which other platforms could implement with a crummy threadpool-based approach.

(I hope that if some Solaris folks show up they will be interested in this ticket.)

Attachments:

* [ecfreactor.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/724/7244a5993b9c6038702545c921538ac438a8ae3c/4b2f4c7ead0d16b6ccfed03e3e349c835dbdc182.patch) (154051 bytes) - added by jvenus on 2012-05-04 19:35:42Z - Native Solaris Event Completion Framework Reactor (RFC Patch)

<details><summary>Searchable metadata</summary>

```
trac-id__5538 5538
type__enhancement enhancement
reporter__glyph glyph
priority__lowest lowest
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__request_for_comment request for comment
time__1331617338000000 1331617338000000
changetime__1341466643000000 1341466643000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jvenus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|jvenus commented|
|-|-|

I registered just because of this ticket.  I am interested in lending a hand for this.  I have access to Solaris sparc/x86 boxes.  I am comfortable with C, Cython, and Twisted.  Let me know where I can help.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Replying to [jvenus](#note_2):
> I registered just because of this ticket.  I am interested in lending a hand for this.  I have access to Solaris sparc/x86 boxes.  I am comfortable with C, Cython, and Twisted.  Let me know where I can help.

Wow, that was quick!  And that sounds awesome.

First: if you haven't already, read the TwistedDevelopment page on the wiki.

There are two areas we could use help.  One is that somebody ultimately needs to write the actual patch - you can just go ahead and do that :).  The other posix-ish reactors, for example, `cfreactor`, or `epollreactor`, will probably demonstrate a reasonable way to integrate a new multiplexing mechanism with our existing socket I/O code.  It's probably less work than it sounds like :).

The other thing that we'd need in order to accept said patch is a Solaris buildslave that could run the tests to make sure it works, and, more importantly, to make sure it keeps working even as people working on other platforms change the code that it depends on.  In order to actually be useful and remain supported, you'd also need to give the Twisted core development team access to the machine running the buildslave, so that we can restart it or inspect it as necessary.  (We don't necessarily need access to the base OS; I'm pretty sure that a restricted zone or jail or otherwise isolated, virtualized environment would be fine, assuming the host environment is reliable enough.)

Thanks a ton for registering your interest, though!

**Comment by twisted-trac** [comment]
|<img alt="jvenus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|jvenus commented|
|-|-|

I have started laying the foundation for this reactor, at the moment it is still in pretty rough shape.  I will periodically push the code to my public github repo and update this ticket as I make progress on it.  Please feel free to comment on it at any point.


https://github.com/JustinVenus/AIOReactor

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Replying to [jvenus](#note_4):
> I have started laying the foundation for this reactor, at the moment it is still in pretty rough shape.  I will periodically push the code to my public github repo and update this ticket as I make progress on it.  Please feel free to comment on it at any point.

> https://github.com/JustinVenus/AIOReactor

Sounds good.  Thanks for looking into this.  Please feel free to ask questions here, on the mailing list, or on IRC as you progress.  Especially, since I notice your code has no tests, feel free to ask questions about how to develop a new reactor in a test-driven way; it will make it a lot easier to get reasonable code coverage and eventually get your code merged into trunk :).  (In the meanwhile, if you want to distribute the reactor separately, there is a plug-in mechanism to register new reactors which you may want to investigate.)

Also: if you like to work using Git, would you mind perhaps helping us out with [our somewhat confused git mirror situation](https://twistedmatrix.com/trac/wiki/GitMirror)?  Lots of people want Twisted to have a presence of some kind on github, but right now [the official github mirror](https://github.com/twisted/twisted) is de-synchronized and not being updated.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

One particular comment about the code there (after having looked for only a minute) is that I'd prefer to see as little Cython as possible.  Compare to epoll reactor, which uses a minimal Cython-implemented epoll library and is otherwise pure Python.  Python is easier to maintain (more people are familiar with it), higher level (eg, no memory management), and more portable (eg, to PyPy).

**Comment by twisted-trac** [comment]
|<img alt="jvenus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|jvenus commented|
|-|-|

Replying to [exarkun](#note_6):
> One particular comment about the code there (after having looked for only a minute) is that I'd prefer to see as little Cython as possible.  Compare to epoll reactor, which uses a minimal Cython-implemented epoll library and is otherwise pure Python.  Python is easier to maintain (more people are familiar with it), higher level (eg, no memory management), and more portable (eg, to PyPy).

I agree the amount of Cython needs to be reduced, and I will do that as I attempt to get this into a working state.  Unfortunately some memory management can't be avoided due to the way the Solaris AIO API works.  Implementation wise this reactor will probably be closer to iocpreactor than it is to epollreactor.

#API Docs
http://docs.oracle.com/cd/E19455-01/806-4750/chap7rt-51/index.html


I would like bring up a limitation, with AIO on solaris, that I have found.  Solaris can only perform 200 AIO operations per process, it is a hard coded limit in one of the header files.  I have some ideas on how to handle this, but this would be a good point for someone other than myself to decide if that is an acceptable limitation.

**Comment by twisted-trac** [comment]
|<img alt="jvenus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|jvenus commented|
|-|-|

Replying to [jvenus](#note_7):
> Replying to [exarkun](#note_6):
> > One particular comment about the code there (after having looked for only a minute) is that I'd prefer to see as little Cython as possible.  Compare to epoll reactor, which uses a minimal Cython-implemented epoll library and is otherwise pure Python.  Python is easier to maintain (more people are familiar with it), higher level (eg, no memory management), and more portable (eg, to PyPy).
> 
> I agree the amount of Cython needs to be reduced, and I will do that as I attempt to get this into a working state.  Unfortunately some memory management can't be avoided due to the way the Solaris AIO API works.  Implementation wise this reactor will probably be closer to iocpreactor than it is to epollreactor.
> 
> #API Docs
> http://docs.oracle.com/cd/E19455-01/806-4750/chap7rt-51/index.html
> 
> 
> I would like bring up a limitation, with AIO on solaris, that I have found.  Solaris can only perform 200 AIO operations per process, it is a hard coded limit in one of the header files.  I have some ideas on how to handle this, but this would be a good point for someone other than myself to decide if that is an acceptable limitation.


#I made an error with my last comment.
"only perform 200 AIO operations per process" should have been written as "ony perform 200 concurrent AIO operations per process"

**Comment by twisted-trac** [comment]
|<img alt="jvenus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|jvenus commented|
|-|-|

I just attached a patch and am looking for feedback.  One Thing I know needs to be done right off of the bat, that I have no experience with, is a unittest for the C Extension.

On my solaris system 3 more unittests than expected failed from Trunk prior to the patch being applied and the same behavior existed after the patch was applied.

Side Note:
From my limited testing using the built-in twistd webserver this reactor is 8x faster than the Epollreactor on the same x86 hardware.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

In the future, when you are looking for feedback, please put the 'review' keyword on the ticket, otherwise it is liable to sit around forever with nobody noticing it :).

**Comment by twisted-trac** [comment]
|<img alt="jvenus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|jvenus commented|
|-|-|

I've decided to attack this problem from two different angles using the same strategy.  I am going to mimic the interfaces for linux epoll and just use the epollreactor as is.  The first way will be to open an enhancement ticket for the python selectmodule.c to implement select.epoll[1] on solaris.  The second way will be to implement the interfaces of twisted.python._epoll.so with another piece of pyrex code.  I think this will minimize the invasive changes and reduce the amount of code necessary to support yet another optimized reactor for a secondary platform.  Thoughts or concerns?

https://github.com/JustinVenus/AIOReactor/blob/master/ecpf_epoll.c
